File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.swift
var input = Int(readLine()!)!
var numbers = [[Int]]()
var minValue = Int.max
var visited = Array(repeating: false, count: input)

for _ in 0..<input {
    numbers.append(readLine()!.split(separator: " ").map {Int(String($0))!})
}

func DFS(depth: Int, idx: Int) {
    if depth == input/2 {
        var sttart = 0
        var link = 0
        for j in 0..<input {
            for k in 0..<input {
                if visited[j] == true && visited[k] == true {
                    sttart += numbers[j][k]
                }
                if visited[j] == false && visited[k] == false {
                    link += numbers[j][k]
                }
            }
        }
        minValue = min(minValue, abs(sttart-link))
        return
    }
    for m in idx..<input {
        if !visited[m] {
            visited[m] = true
            DFS(depth: depth+1, idx: m)
            visited[m] = false
        }
    }
}

DFS(depth: 0, idx: 0)
print(minValue)

File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.py
number = int(input())
board = []
answer = int(1e9)
for _ in range(number):
    board.append(list(map(int, input().split())))
visited = [False for _ in range(number)]
def DFS(depth, start):
    global answer
    if depth == number // 2:
        A = 0
        B = 0

        for i in range(number):
            for j in range(number):
                if visited[i] and visited[j]:
                    A += board[i][j]
                if not visited[i] and not visited[j]:
                    B += board[i][j]
        answer = min(answer, abs(A-B))
        return

    for v in range(start, number):
        visited[v] = True
        DFS(depth+1, v+1)
        visited[v] = False

DFS(0, 0)
print(answer)